base/strings/util.jl

```diff
@@ -633,30 +633,36 @@ julia> hex2bytes(a)
 function hex2bytes end
 
 hex2bytes(s::AbstractString) = hex2bytes(String(s))
-hex2bytes(s::Union{String,AbstractVector{UInt8}}) = hex2bytes!(Vector{UInt8}(undef, length(s) >> 1), s)
-
-_firstbyteidx(s::String) = 1
-_firstbyteidx(s::AbstractVector{UInt8}) = first(eachindex(s))
-_lastbyteidx(s::String) = sizeof(s)
-_lastbyteidx(s::AbstractVector{UInt8}) = lastindex(s)
+hex2bytes(s::String) = hex2bytes(transcode(UInt8, s))
+hex2bytes(s) = hex2bytes!(Vector{UInt8}(undef, length(s) >> 1), s)
 
 """
-    hex2bytes!(d::AbstractVector{UInt8}, s::Union{String,AbstractVector{UInt8}})
+    hex2bytes!(dest::AbstractVector{UInt8}, itr)
 
-Convert an array `s` of bytes representing a hexadecimal string to its binary
+Convert an iterable `itr` of bytes representing a hexadecimal string to its binary
 representation, similar to [`hex2bytes`](@ref) except that the output is written in-place
-in `d`.   The length of `s` must be exactly twice the length of `d`.
-"""
-function hex2bytes!(d::AbstractVector{UInt8}, s::Union{String,AbstractVector{UInt8}})
-    if 2length(d) != sizeof(s)
-        isodd(sizeof(s)) && throw(ArgumentError("input hex array must have even length"))
-        throw(ArgumentError("output array must be half length of input array"))
+to `dest`. The length of `dest` must be at least half the length of `itr`.
+"""
+function hex2bytes!(dest::AbstractArray{UInt8}, itr)
+    isodd(length(itr)) && throw(ArgumentError("length of iterable must be even"))
+    @boundscheck 2*length(dest) < length(itr) && throw(ArgumentError("length of output array must be at least half of the length of input array"))
+    iszero(length(itr)) && return dest
+
+    # we know these iterations always work because of the checks above
+    # as a bonus, we don't have to check the result of `iterate` for `nothing`!
+    # unfortunately, we have to do them explicitly here because the first `iterate` has a different signature
+    (x,state) = iterate(itr)
+    (y,state) = iterate(itr, state)
+    dest[firstindex(dest)] = number_from_hex(x) << 4 + number_from_hex(y)
+
+    # incorporating the iterations into this loop via Iterators.partition was slower than the original function
+    @inbounds for i in Iterators.drop(eachindex(dest),1)
+        (x,state) = iterate(itr, state)
+        (y,state) = iterate(itr, state)
+        dest[i] = number_from_hex(x) << 4 + number_from_hex(y)
```

Comment:
(I feel like this is a premature optimization, though...)
In general, it is more compiler-friendly to branch by explicitly comparing the iterator result with `nothing`.
For example:
```julia
    next = iterate(itr)
    @inbounds for i in eachindex(dest)
        next === nothing && break
        x, state = next
        next = iterate(itr, state)
        next === nothing && break
        y, state = next
        next = iterate(itr, state)
        dest[i] = number_from_hex(x) << 0x4 + number_from_hex(y)
    end
```
Of course, the final decision should be made based on benchmarks.

Reply:
You're right, I see an improvement of +16% across the board by explicitly checking for `nothing` 🤔 I'm not sure I understand why though. From my understanding, the length check above should be enough to prove this property, right?  Further, if I now use the `nfh` function from my comment above instead of the branching version, I get an additional ~20% improvement when the argument passed is `String`!

<details>
<summary>Benchmarking numbers for explicit `nothing` checks and linearized `number_from_hex`</summary>

```
minimum:
6-element BenchmarkTools.BenchmarkGroup:
  tags: []
  10000 => TrialJudgement(-36.35% => improvement)
  100000 => TrialJudgement(-37.31% => improvement)
  1000000 => TrialJudgement(-37.09% => improvement)
  10000000 => TrialJudgement(-34.46% => improvement)
  1000 => TrialJudgement(-31.17% => improvement)
  100 => TrialJudgement(-10.53% => improvement)

mean:
6-element BenchmarkTools.BenchmarkGroup:
  tags: []
  10000 => TrialJudgement(-36.16% => improvement)
  100000 => TrialJudgement(-37.88% => improvement)
  1000000 => TrialJudgement(-37.19% => improvement)
  10000000 => TrialJudgement(-34.95% => improvement)
  1000 => TrialJudgement(-27.14% => improvement)
  100 => TrialJudgement(-14.40% => improvement)

median:
6-element BenchmarkTools.BenchmarkGroup:
  tags: []
  10000 => TrialJudgement(-36.98% => improvement)
  100000 => TrialJudgement(-40.32% => improvement)
  1000000 => TrialJudgement(-37.30% => improvement)
  10000000 => TrialJudgement(-34.86% => improvement)
  1000 => TrialJudgement(-31.33% => improvement)
  100 => TrialJudgement(-14.29% => improvement)

maximum:
6-element BenchmarkTools.BenchmarkGroup:
  tags: []
  10000 => TrialJudgement(-3.05% => invariant)
  100000 => TrialJudgement(-11.49% => improvement)
  1000000 => TrialJudgement(-35.83% => improvement)
  10000000 => TrialJudgement(-36.27% => improvement)
  1000 => TrialJudgement(-13.45% => improvement)
  100 => TrialJudgement(+1.94% => invariant)
```

</details>


I did benchmark the original version against the one I first wrote here to make sure there's no regression introduced by generalizing the code.

Reply:
Ok, there's gotta be something else going on here - if I use the version without explicit `nothing` checks but _with_ `nfh` from above, I also get the -30% improvement over the original function 🤔

Reply:
I.e. this

<details>
<summary>code</summary>

```
h2b(s::AbstractString) = h2b(String(s))
h2b(s::String) = h2b(transcode(UInt8, s))
h2b(s) = h2b!(Vector{UInt8}(undef, length(s) >> 1), s)

function h2b!(dest::AbstractVector{UInt8}, itr)
    isodd(length(itr)) && throw(ArgumentError("length of iterable must be even"))
    @boundscheck 2*length(dest) < length(itr) && throw(ArgumentError("length of output array must be at least half of the length of input array"))
    iszero(length(itr)) && return dest

    next = iterate(itr)
    @inbounds for i in eachindex(dest)
        x,state = next
        y,state = iterate(itr, state)
        next = iterate(itr, state)
        dest[i] = nfh(x) << 4 + nfh(y)
    end

    return dest
end

@inline nfh(c::Char) = nfh(UInt8(c))
@inline function nfh(c::UInt8)
    ret = (c >= UInt8('0')) & (c <= UInt8('9'))
    c |= 0b0100000
    ret |= (c > UInt8('`')) & (c <= UInt8('f'))
    !ret && throw(ArgumentError("byte is not an ASCII hexadecimal digit"))
    return c - (c & 0xf1) - (c & 0x10 ) * 0x0a
end
```
</details>

benchmarked like

<details>
<summary>benchmark</summary>

```
julia> old = BenchmarkGroup()

julia> new = BenchmarkGroup()

julia> for i in [100, 1_000, 10_000, 100_000, 1_000_000, 10_000_000]
         old[i] = @benchmarkable hex2bytes(str) setup=(str=randstring("0123456789abcdefABCDEF", $i)) evals=1
         new[i] = @benchmarkable h2b(str) setup=(str=randstring("0123456789abcdefABCDEF", $i)) evals=1
       end

julia> new_res, old_res = run(new), run(old); for f in (minimum, mean, median, maximum)
        println("\n$f:"); display(judge(f(new_res), f(old_res)))
       end
```
</details>

leads to 

<details>
<summary>results</summary>

```
minimum:
6-element BenchmarkTools.BenchmarkGroup:
  tags: []
  10000 => TrialJudgement(-22.73% => improvement)
  100000 => TrialJudgement(-23.87% => improvement)
  1000000 => TrialJudgement(-23.90% => improvement)
  10000000 => TrialJudgement(-21.91% => improvement)
  1000 => TrialJudgement(-19.48% => improvement)
  100 => TrialJudgement(-5.26% => improvement)

mean:
6-element BenchmarkTools.BenchmarkGroup:
  tags: []
  10000 => TrialJudgement(-18.44% => improvement)
  100000 => TrialJudgement(-24.28% => improvement)
  1000000 => TrialJudgement(-23.71% => improvement)
  10000000 => TrialJudgement(-20.84% => improvement)
  1000 => TrialJudgement(-18.86% => improvement)
  100 => TrialJudgement(-9.36% => improvement)

median:
6-element BenchmarkTools.BenchmarkGroup:
  tags: []
  10000 => TrialJudgement(-23.76% => improvement)
  100000 => TrialJudgement(-24.43% => improvement)
  1000000 => TrialJudgement(-23.44% => improvement)
  10000000 => TrialJudgement(-22.29% => improvement)
  1000 => TrialJudgement(-20.48% => improvement)
  100 => TrialJudgement(-9.52% => improvement)

maximum:
6-element BenchmarkTools.BenchmarkGroup:
  tags: []
  10000 => TrialJudgement(+2183.23% => regression)
  100000 => TrialJudgement(-60.49% => improvement)
  1000000 => TrialJudgement(-39.49% => improvement)
  10000000 => TrialJudgement(-10.78% => improvement)
  1000 => TrialJudgement(-36.09% => improvement)
  100 => TrialJudgement(+506.73% => regression)
```
</details>

Reply:
This will not be SIMD vectorized, so the `c |= 0b0100000` hack should be enough. Of course, it can be SIMD-optimized if both inputs are `Vector`s, etc., but that's obviously a premature optimization.
```julia
function nfh(c::UInt8)
    UInt8('0') <= c <= UInt8('9') && return c - UInt8('0') 
    c |= 0b0100000
    UInt8('a') <= c <= UInt8('f') && return c - UInt8('a') + 0x0a 
    throw(ArgumentError("byte is not an ASCII hexadecimal digit"))
end
```

Reply:
This improves the average case quite a bit, giving an additional -3-4% improvement, for a total of ~-34% compared to the original implementation when the input is always hitting `return` instead of `throw` :) I did not think the branch was this cheap here.

I'll incorporate it.